### ISSUE TITLE
issue running plot_read_length_by_novelty

### ISSUE THREAD

**Issue by callumparr** [mortazavilab/TALON-paper-2020#3]
python version 3.6.7, also tried python 3.7.2

```
python TALON-paper-2020/plotting_scripts/plot_read_length_by_novelty.py --f Mouse_aging/TALON/TALON_dRNA_mm10_talon_read_annot_mod.tsv --platform MinION-dRNA --dataset M_SPF_old_rep1,M_SPF_old_rep2 --whitelist $pb_whitelist --o Mouse_aging/TALON/plots/length
```

Error 

```
sys:1: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
Traceback (most recent call last):
  File "TALON-paper-2020/plotting_scripts/plot_read_length_by_novelty.py", line 131, in <module>
    main()
  File "TALON-paper-2020/plotting_scripts/plot_read_length_by_novelty.py", line 114, in main
    data = data[~filter]
  File "/home/callum/.local/lib/python3.7/site-packages/pandas/core/generic.py", line 1324, in __invert__
    new_data = self._mgr.apply(operator.invert)
  File "/home/callum/.local/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 407, in apply
    applied = b.apply(f, **kwargs)
  File "/home/callum/.local/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 346, in apply
    result = func(self.values, **kwargs)
TypeError: bad operand type for unary ~: 'float'
```

Input files 

annot file

```
read_name	dataset	genome_build	chrom	read_start	read_end	strand	n_exons	read_length	gene_ID	transcript_ID	annot_gene_id	annot_transcript_id	annot_gene_name	annot_transcript_name	gene_novelty	transcript_novelty	ISM_subtype	fraction_As	custom_label	allelic_label	start_support	end_support
6eebcb2a-b5d1-40da-b409-35990366af14	M_GF_old_rep1	mus_musculus_99	16	3292339	3292512	+	1	174	55472	142648	TALONG000055472	TALONT000142648	TALONG000055472	TALONT000142648	Intergenic	Intergenic	None	0.44999998807907104	None	None	None	None
064b1390-57ba-40d4-a9a9-ee64906f7d6c	F_SPF_old_rep4	mus_musculus_99	12	3063154	3063323	+	1	170	55473	142649	TALONG000055473	TALONT000142649	TALONG000055473	TALONT000142649	Intergenic	Intergenic	None	0.20000000298023224	None	None	None	None
eb461894-cf1a-4942-a333-8287b8321cfe	F_GF_old_rep4	mus_musculus_99	17	3044756	3043851	-	2	751	48667	142650	ENSMUSG00000068141	TALONT000142650	Gm10232	TALONT000142650	Known	Genomic	None	0.25	None	None	None	None
8560943d-d013-47fd-906a-5aacbf0824fa	F_SPF_young_rep5	mus_musculus_99	19	3033983	3034426	+	1	444	55474	142651	TALONG000055474	TALONT000142651	TALONG000055474	TALONT000142651	Intergenic	IntergeniNone	0.3499999940395355	None	None	None	None
e4ffa640-7b20-40e3-8c04-0a8c311264f0	M_SPF_young_rep3	mus_musculus_99	14	3049085	3076842	+	7	2182	31859	81816	ENSMUSG00000079410	ENSMUST00000166494	Gm2897	Gm2897-201	Known	Known	None	0.15000000596046448	None	None	None	None
e1830ba0-c4a9-4ea5-a744-0f8758683071	M_SPF_young_rep2	mus_musculus_99	6	3185494	3184697	-	1	798	55475	142652	TALONG000055475	TALONT000142652	TALONG000055475	TALONT000142652	Intergenic	IntergeniNone	0.0	None	None	None	None
2655636e-1a07-453d-8902-1f233593d20d	F_GF_old_rep4	mus_musculus_99	17	3044254	3043852	-	1	390	48667	125223	ENSMUSG00000068141	ENSMUST00000089221	Gm10232	Gm10232-201	Known	Known	None	0.30000001192092896	None	None	None	None
23d152b9-9020-4182-aead-47bf81d15779	M_GF_old_rep3	mus_musculus_99	17	3044748	3043858	-	1	891	48667	125223	ENSMUSG00000068141	ENSMUST00000089221	Gm10232	Gm10232-201	Known	Known	None	0.44999998807907104	None	None	None	None
d48ab7dd-4440-4f17-aa74-bc0dd44c2977	M_GF_old_rep4	mus_musculus_99	13	3100107	3096030	-	1	4070	55476	142653	TALONG000055476	TALONT000142653	TALONG000055476	TALONT000142653	Antisense	Antisense	None	0.05000000074505806	None	None	None	None

```

whitelist
```
10000,24918
10002,24920
10003,24923
10004,24925
10005,184781
10005,190747
10005,191306
10005,191311
10005,191320
10005,191757
```

**Comment by callumparr** [comment]
I check the data types in columns in python and it outputs the same as the example read annotation file from TALON-5.0 repository. 

```
>>> import pandas as pd
>>> data = pd.read_csv("Mouse_aging/TALON/TALON_dRNA_mm10_talon_read_annot_mod.tsv", sep = "\t", header = 0)
sys:1: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
>>> data.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45496774 entries, 0 to 45496773
Data columns (total 23 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   read_name              object 
 1   dataset                object 
 2   genome_build           object 
 3   chrom                  object 
 4   read_start             int64  
 5   read_end               int64  
 6   strand                 object 
 7   n_exons                int64  
 8   read_length            int64  
 9   gene_ID                int64  
 10  transcript_ID          int64  
 11  annot_gene_id          object 
 12  annot_transcript_id    object 
 13  annot_gene_name        object 
 14  annot_transcript_name  object 
 15  gene_novelty           object 
 16  transcript_novelty     object 
 17  ISM_subtype            object 
 18  fraction_As            float64
 19  custom_label           object 
 20  allelic_label          object 
 21  start_support          object 
 22  end_support            object 
dtypes: float64(1), int64(6), object(16)
memory usage: 7.8+ GB

```

**Comment by fairliereese** [comment]
Hi, from Googling your error, it looks like you might somehow have NaNs in your index column. (Source: https://stackoverflow.com/questions/52297740/typeerror-bad-operand-type-for-unary-float). Can you see if this is the case? You can try adding the `na=False` option to the `contains` call on lines 113 and 115. If that fixes it then you probably have NaNs somewhere, which I'm not sure I know where they would have come from.

**Comment by callumparr** [comment]
Thank you for your reply and your help.

I worried that I originally had to alter the reference (taken from ensembl (`Genome_files/Mus_musculus.GRCm38.dna.primary_assembly.fa`)  . because it had various additions to the basic chromosome name. https://github.com/mortazavilab/TranscriptClean/issues/23#issue-906491161


```
❯ grep ">" Genome_files/Mus_musculus.GRCm38.dna.primary_assembly_cleanheader.fa
>1
>10
>11
>12
>13
>14
>15
>16
>17
>18
>19
>2
>3
>4
>5
>6
>7
>8
>9
>MT
>X
>Y
>JH584299.1
>GL456233.1
>JH584301.1
>GL456211.1
>GL456350.1
>JH584293.1
>GL456221.1
>JH584297.1
>JH584296.1
>GL456354.1
>JH584294.1
>JH584298.1
>JH584300.1
>GL456219.1
>GL456210.1
>JH584303.1
>JH584302.1
>GL456212.1
>JH584304.1
>GL456379.1
>GL456216.1
>GL456393.1
>GL456366.1
>GL456367.1
>GL456239.1
>GL456213.1
>GL456383.1
>GL456385.1
>GL456360.1
>GL456378.1
>GL456389.1
>GL456372.1
>GL456370.1
>GL456381.1
>GL456387.1
>GL456390.1
>GL456394.1
>GL456392.1
>GL456382.1
>GL456359.1
>GL456396.1
>GL456368.1
>JH584292.1
>JH584295.1
```

from original

```
 grep ">" Genome_files/Mus_musculus.GRCm38.dna.primary_assembly.fa
>1 dna:chromosome chromosome:GRCm38:1:1:195471971:1 REF
>10 dna:chromosome chromosome:GRCm38:10:1:130694993:1 REF
>11 dna:chromosome chromosome:GRCm38:11:1:122082543:1 REF
>12 dna:chromosome chromosome:GRCm38:12:1:120129022:1 REF
>13 dna:chromosome chromosome:GRCm38:13:1:120421639:1 REF
>14 dna:chromosome chromosome:GRCm38:14:1:124902244:1 REF
>15 dna:chromosome chromosome:GRCm38:15:1:104043685:1 REF
>16 dna:chromosome chromosome:GRCm38:16:1:98207768:1 REF
>17 dna:chromosome chromosome:GRCm38:17:1:94987271:1 REF
>18 dna:chromosome chromosome:GRCm38:18:1:90702639:1 REF
>19 dna:chromosome chromosome:GRCm38:19:1:61431566:1 REF
>2 dna:chromosome chromosome:GRCm38:2:1:182113224:1 REF
>3 dna:chromosome chromosome:GRCm38:3:1:160039680:1 REF
>4 dna:chromosome chromosome:GRCm38:4:1:156508116:1 REF
>5 dna:chromosome chromosome:GRCm38:5:1:151834684:1 REF
>6 dna:chromosome chromosome:GRCm38:6:1:149736546:1 REF
>7 dna:chromosome chromosome:GRCm38:7:1:145441459:1 REF
>8 dna:chromosome chromosome:GRCm38:8:1:129401213:1 REF
>9 dna:chromosome chromosome:GRCm38:9:1:124595110:1 REF
>MT dna:chromosome chromosome:GRCm38:MT:1:16299:1 REF
>X dna:chromosome chromosome:GRCm38:X:1:171031299:1 REF
>Y dna:chromosome chromosome:GRCm38:Y:1:91744698:1 REF
>JH584299.1 dna:scaffold scaffold:GRCm38:JH584299.1:1:953012:1 REF
>GL456233.1 dna:scaffold scaffold:GRCm38:GL456233.1:1:336933:1 REF
>JH584301.1 dna:scaffold scaffold:GRCm38:JH584301.1:1:259875:1 REF
>GL456211.1 dna:scaffold scaffold:GRCm38:GL456211.1:1:241735:1 REF
>GL456350.1 dna:scaffold scaffold:GRCm38:GL456350.1:1:227966:1 REF
>JH584293.1 dna:scaffold scaffold:GRCm38:JH584293.1:1:207968:1 REF
>GL456221.1 dna:scaffold scaffold:GRCm38:GL456221.1:1:206961:1 REF
>JH584297.1 dna:scaffold scaffold:GRCm38:JH584297.1:1:205776:1 REF
>JH584296.1 dna:scaffold scaffold:GRCm38:JH584296.1:1:199368:1 REF
>GL456354.1 dna:scaffold scaffold:GRCm38:GL456354.1:1:195993:1 REF
>JH584294.1 dna:scaffold scaffold:GRCm38:JH584294.1:1:191905:1 REF
>JH584298.1 dna:scaffold scaffold:GRCm38:JH584298.1:1:184189:1 REF
>JH584300.1 dna:scaffold scaffold:GRCm38:JH584300.1:1:182347:1 REF
>GL456219.1 dna:scaffold scaffold:GRCm38:GL456219.1:1:175968:1 REF
>GL456210.1 dna:scaffold scaffold:GRCm38:GL456210.1:1:169725:1 REF
>JH584303.1 dna:scaffold scaffold:GRCm38:JH584303.1:1:158099:1 REF
>JH584302.1 dna:scaffold scaffold:GRCm38:JH584302.1:1:155838:1 REF
>GL456212.1 dna:scaffold scaffold:GRCm38:GL456212.1:1:153618:1 REF
>JH584304.1 dna:scaffold scaffold:GRCm38:JH584304.1:1:114452:1 REF
>GL456379.1 dna:scaffold scaffold:GRCm38:GL456379.1:1:72385:1 REF
>GL456216.1 dna:scaffold scaffold:GRCm38:GL456216.1:1:66673:1 REF
>GL456393.1 dna:scaffold scaffold:GRCm38:GL456393.1:1:55711:1 REF
>GL456366.1 dna:scaffold scaffold:GRCm38:GL456366.1:1:47073:1 REF
>GL456367.1 dna:scaffold scaffold:GRCm38:GL456367.1:1:42057:1 REF
>GL456239.1 dna:scaffold scaffold:GRCm38:GL456239.1:1:40056:1 REF
>GL456213.1 dna:scaffold scaffold:GRCm38:GL456213.1:1:39340:1 REF
>GL456383.1 dna:scaffold scaffold:GRCm38:GL456383.1:1:38659:1 REF
>GL456385.1 dna:scaffold scaffold:GRCm38:GL456385.1:1:35240:1 REF
>GL456360.1 dna:scaffold scaffold:GRCm38:GL456360.1:1:31704:1 REF
>GL456378.1 dna:scaffold scaffold:GRCm38:GL456378.1:1:31602:1 REF
>GL456389.1 dna:scaffold scaffold:GRCm38:GL456389.1:1:28772:1 REF
>GL456372.1 dna:scaffold scaffold:GRCm38:GL456372.1:1:28664:1 REF
>GL456370.1 dna:scaffold scaffold:GRCm38:GL456370.1:1:26764:1 REF
>GL456381.1 dna:scaffold scaffold:GRCm38:GL456381.1:1:25871:1 REF
>GL456387.1 dna:scaffold scaffold:GRCm38:GL456387.1:1:24685:1 REF
>GL456390.1 dna:scaffold scaffold:GRCm38:GL456390.1:1:24668:1 REF
>GL456394.1 dna:scaffold scaffold:GRCm38:GL456394.1:1:24323:1 REF
>GL456392.1 dna:scaffold scaffold:GRCm38:GL456392.1:1:23629:1 REF
>GL456382.1 dna:scaffold scaffold:GRCm38:GL456382.1:1:23158:1 REF
>GL456359.1 dna:scaffold scaffold:GRCm38:GL456359.1:1:22974:1 REF
>GL456396.1 dna:scaffold scaffold:GRCm38:GL456396.1:1:21240:1 REF
>GL456368.1 dna:scaffold scaffold:GRCm38:GL456368.1:1:20208:1 REF
>JH584292.1 dna:scaffold scaffold:GRCm38:JH584292.1:1:14945:1 REF
>JH584295.1 dna:scaffold scaffold:GRCm38:JH584295.1:1:1976:1 REF
```

I tried checking the column of the index to see if any NaN but cannot see there is anything like this in the read annotation file.

In any case, by making the changes in the plotting script that you suggested I could get an output plot

```
❯ python TALON-paper-2020/plotting_scripts/plot_read_length_by_novelty.py --f Mouse_aging/TALON/TALON_dRNA_mm10_talon_read_annot_mod.tsv --datasets M_SPF_old_rep1,M_SPF_old_rep2,M_SPF_old_rep3,M_SPF_old_rep4 --platform MinION-dRNA --whitelist $pb_whitelist --o Mouse_aging/TALON/plots/length --ymax 10000
sys:1: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
transcript_novelty
Known          80.0
ISM            85.0
NIC           186.0
NNC           139.0
Antisense      98.0
Intergenic     84.0
Name: read_length, dtype: float64
transcript_novelty
Known         15800.0
ISM           15311.0
NIC           10487.0
NNC            8062.0
Antisense     16030.0
Intergenic     8524.0
Name: read_length, dtype: float64
No handles with labels found to put in legend. 
```

The other issue I had was that I tried to run on all of the datasets but not flagging a particular dataset but it would not run like this. Is it expected to provide a comma separate list of datasets because from the help it says the default behaviour is to use all of the datasets from the annotation file?

I can just list all of the datasets to get around this. I also plan to run iteratively for each condition as well.

**Comment by callumparr** [comment]
Reading that stackoverflow seems any empty space or NaN but far as I can tell there are no such values in my data frame. 
It is running now anyway so I think I can ignore it. 
```
data.is.null().values.any()
false
```

**Comment by fairliereese** [comment]
> The other issue I had was that I tried to run on all of the datasets but not flagging a particular dataset but it would not run like this. Is it expected to provide a comma separate list of datasets because from the help it says the default behaviour is to use all of the datasets from the annotation file?
> 
> I can just list all of the datasets to get around this. I also plan to run iteratively for each condition as well.

Hmm, it should work for all datasets, I would be able to help you more if you included your error message you're getting when trying that though.

**Comment by callumparr** [comment]
```
python TALON-paper-2020/plotting_scripts/plot_read_length_by_novelty.py --f Mouse_aging/TALON/TALON_dRNA_mm10_talon_read_annot_mod.tsv --platform MinION-dRNA --whitelist $whitelist --o Mouse_aging/TALON/plots/length/ --ymax 10000
```

If I want it to run on all datasets I was thinking not to use `--datasets`.

But in this case get the error.

```
sys:1: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
Traceback (most recent call last):
  File "TALON-paper-2020/plotting_scripts/plot_read_length_by_novelty.py", line 131, in <module>
    main()
  File "TALON-paper-2020/plotting_scripts/plot_read_length_by_novelty.py", line 125, in main
    fname = options.outprefix + "_".join(datasets) + options.platform + \
UnboundLocalError: local variable 'datasets' referenced before assignment
```

If I try putting all the names of datasets (around 34) with --dataset the plot can not write out as the filename will be too long. Although this is probably easy enough to solve by editing the script to provide a specific output filename explicitly.

**Comment by callumparr** [comment]
Also, the y-axes plot of the number of reads binned to distance from TES seems to not be determined by the --ymax as the plot always has a maximum y value of 250,000. Is there a way to get this to scale appropriately to the highest peak?

**Comment by fairliereese** [comment]
The datasets thing should work now if you redownload the script. I should have fixed the bug. And it actually looks like the ymax argument is never used and the ylimit is just automatically being calculated. If none of your data is being cut off I wouldn't worry about it too much.

**Comment by callumparr** [comment]
Great thank you. I think most of the changes I've made are very specific to my environment and/or the read annotation file I generated. In particular, any script that is using the 'chrom' column. But if it's generally helpful to others can generate a PR. 

https://github.com/callumparr/TALON-paper-2020/commit/ad87c22acb018f8cbbf96ee039f9e041ac7c6cd4#diff-ac3fcf2ba164f5fe96d3721b49ad2291a5d69a02c2097c251caa11747168741c
https://github.com/callumparr/TALON-paper-2020/commit/6bb288caf0e5b38d23f8e8f7b1123f04ff0e187b
